dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,193 @@ public Pattern getParameterPattern() {
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+                default:
+                    throw new SQLException("Unpexpected procedure parameter type for parameter " +
+                                  rs.getString("COLUMN_NAME"));
+            }
+        }
+    }
+
+    private class DatabaseFunction extends DatabaseObject {
+
+        private DatabaseFunction(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        public boolean exists() throws SQLException {
+            try (ResultSet rs = getObjects()) {
+                return rs.next();
+            }
+        }
+
+        @Override
+        protected ResultSet getObjects() throws SQLException {
+            return getMetaData().getFunctions(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getMetaData().getFunctionColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            int paramType = rs.getInt("COLUMN_TYPE");
+            switch (paramType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+                default:
+                    throw new SQLException("Unpexpected function parameter type for parameter " +
+                                  rs.getString("COLUMN_NAME"));
+            }
+        }
+    }
+
+    private class DatabaseObjectName {
+        String schemaName, objectName;
+
+        private DatabaseObjectName(String defaultSchemaName, String objName) {
+            String[] qualifiers = objName.split("\\.");
+            schemaName = (qualifiers.length == 2) ? qualifiers[0] : defaultSchemaName;
+            objectName = (qualifiers.length == 2) ? qualifiers[1] : qualifiers[0];
+        }
+
+        private String getSchemaName() {
+            return schemaName;
+        }
+
+        String getObjectName() {
+            return objectName;
+        }
+    }
+
+    public Map<String, DbParameterAccessor> getAllProcedureParameters(String callableName)
+                                                                       throws SQLException {
+        String name = callableName.toUpperCase();
+        DatabaseObjectName objName = new DatabaseObjectName(getConnection().getSchema(), name);
+        DatabaseProcedure dbproc = new DatabaseProcedure(objName);
+        if (dbproc.exists()) {
+            return dbproc.getParams();
+        } else {
+            DatabaseFunction dbfunc = new DatabaseFunction(objName);
```

Comment:
Hmm, that still looks a bit complicated if compared to the other. And indeed with this design of `exists` it's not quite the same as it used to be in previous PR. What about having one more helper method to get the object. E.g. something like:
``` java
DatabaseObject getStoredCall(objName) {
    List<DatabaseObject> prioritisedCandidates = Arrays.asList(
        new DatabaseProcedure(objName),
        new DatabaseFunction(objName));

    for (DatabaseObject object : prioritisedCandidaes)
        if (object.exists()) {
            return object;
        }
    }

    throw new SQLException("Procedure/function " + objName + " does not exist");
}
```